src/rmc.rs

```diff
@@ -0,0 +1,302 @@
+/// multisignature under the hash
+/// * in the multisignature aggregation phase, we store the complete multisignature of the hash
+/// and the set of nodes about who we know that they have a multisignature
+/// * in the completed phase we only store the multisigned hash.
+enum HashState<'a, H: Signable, MK: MultiKeychain> {
+    SignatureAggregation {
+        partial: PartiallyMultisigned<'a, H, MK>,
+    },
+    Completed {
+        complete: Multisigned<'a, H, MK>,
+    },
+}
+
+impl<'a, H: Signable + Clone, MK: MultiKeychain> HashState<'a, H, MK> {
+    fn handle_unchecked_message(
+        maybe_state: Option<Self>,
+        message: Message<H, MK::Signature, MK::PartialMultisignature>,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::Completed { .. }) => maybe_state,
+            _ => match message {
+                Message::MultisignedHash(unchecked) => match unchecked.check_multi(keychain) {
+                    Ok(complete) => Some(HashState::Completed { complete }),
+                    Err(_) => {
+                        debug!(target: "rmc", "Received a hash with a bad signature");
+                        maybe_state
+                    }
+                },
+                Message::SignedHash(unchecked) => {
+                    let signed_hash = match unchecked.check(keychain) {
+                        Ok(signed_hash) => signed_hash,
+                        Err(_) => {
+                            debug!(target: "rmc", "Received a hash with a bad multisignature");
+                            return maybe_state;
+                        }
+                    };
+                    let partial = match maybe_state {
+                        None => signed_hash.into_partially_multisigned(keychain),
+                        Some(HashState::SignatureAggregation { mut partial }) => {
+                            partial.add_signature(signed_hash);
+                            partial
+                        }
+                        Some(HashState::Completed { .. }) => unreachable!(),
+                    };
+                    match partial.try_into_complete(keychain) {
+                        Ok(complete) => {
+                            multisigned_hashes_tx
+                                .unbounded_send(complete.clone())
+                                .expect("sending multisigned hash should succeed");
+                            Some(HashState::Completed { complete })
+                        }
+                        Err(e) => Some(HashState::SignatureAggregation { partial: e.partial }),
+                    }
+                }
+            },
+        }
+    }
+}
+
+pub enum Task<H: Signable, MK: MultiKeychain> {
+    BroadcastMessage(Message<H, MK::Signature, MK::PartialMultisignature>),
+}
+
+#[async_trait]
+pub trait TaskScheduler<T> {
```

Comment:
After Damian's changes this should probably be based on a `DelaySchedule`, but I guess this can be left for the refactoring that unifies the usage of both throughout rush...

Reply:
We talked about this a bit. I think it's good for now and we can try to unify it later. I'm thinking of TaskScheduler that supports DelaySchedules :), but again -- would leave it for later.

Reply:
Correct